### PR TITLE
 The debug flags do not work correctly with headless templates

### PR DESCRIPTION
**Pull request by xm1k3**
## Proposed changes

- `nuclei -t ./headless.yaml -u {URL} -headless -debug / -debug-req / -debug-resp`

```

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.4-dev

		projectdiscovery.io

[INF] Using Nuclei Engine 2.8.4-dev (development)
[INF] Using Nuclei Templates 9.3.1 (latest)
[INF] Templates added in last update: 2
[INF] Templates loaded for scan: 1
[INF] Targets loaded for scan: 1
[INF] [headless] Dumped Headless request for https://twitter.com
[DBG] 	navigate url:https://twitter.com
	waitload
[INF] No results found. Better luck next time!
```

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: v2/pkg/protocols/headless/request.go [hunk 1]
```diff
@@ -98,14 +98,19 @@ func (request *Request) executeRequestWithPayloads(inputURL string, payloads map
...
 
 	reqBuilder := &strings.Builder{}
-	if request.options.Options.Debug || request.options.Options.DebugRequests {
+	if request.options.Options.Debug || request.options.Options.DebugRequests || request.options.Options.DebugResponse {
 		gologger.Info().Msgf("[%s] Dumped Headless request for %s", request.options.TemplateID, inputURL)
 
 		for _, act := range request.Steps {
-			reqBuilder.WriteString(act.String())
+			actStepStr := act.String()
+			if strings.Contains(actStepStr, "{{BaseURL}}") {
```
Comment (Mzack9999) [inline v2/pkg/protocols/headless/request.go:106]: We can replace directly, as internally `strings.ReplaceAll` already uses `strings.Index`:
```go
actStepStr = strings.ReplaceAll(actStepStr, "{{BaseURL}}", inputURL)
```

File: v2/pkg/protocols/headless/request.go [hunk 2]
```diff
@@ -98,14 +98,19 @@ func (request *Request) executeRequestWithPayloads(inputURL string, payloads map
...
+				actStepStr = strings.ReplaceAll(actStepStr, "{{BaseURL}}", inputURL)
+			}
+			reqBuilder.WriteString("\t" + actStepStr)
```
Comment (Mzack9999) [inline v2/pkg/protocols/headless/request.go:109]: What about:
```go
reqBuilder.WriteString("\t" + actStepStr + "\n")
```

**Review by Mzack9999 — CHANGES_REQUESTED** [review]
suggesting small changes

**Review by Mzack9999 — APPROVED** [review]
lgtm

**Review by ehsandeep — APPROVED** [review]
@xm1k3 @Mzack9999 this part can reworked in the future to make it dynamic mapping to replace with supported variables, surprisingly looks like `BaseURL` is only variable supported now in headless.

@Mzack9999 once confirmed we can create a follow-up issue to support existing global variables in headless protocol.

**Comment by xm1k3** [comment]
@ehsandeep yes, perhaps a good idea would be to use some sort of whitelist to manage all the keywords we need to replace
what do you think?

**Pull request merged**